### PR TITLE
topdown/strings: make strings.replace_n use builtins.ObjectOperand

### LINKED ISSUE #2822: builtin `strings.replace_n` is nondetermistic when patterns overlap
I believe this has been noted in #1825, but got lost along the way: since map iteration is nondetermistic, so is the ordering in the replacement array [built here](https://github.com/open-policy-agent/opa/blob/4671822476b0083fc9f539f28db18b527c023d21/topdown/strings.go#L251-L258). Since in `strings.Replacer`, _"old string comparisons are done in argument order"_, this has implications on what gets replaced.

AFAICT it only affects overlapping matches.

## Expected Behavior

No matter how often I run this,
```rego
strings.replace_n({"f": "1", "foo":"2"}, "foo")
```

I always get the same result.

## Actual Behavior

```interactive
> strings.replace_n({"f": "1", "foo":"2"}, "foo")
"1oo"
> strings.replace_n({"f": "1", "foo":"2"}, "foo")
"1oo"
> strings.replace_n({"f": "1", "foo":"2"}, "foo")
"1oo"
> strings.replace_n({"f": "1", "foo":"2"}, "foo")
"1oo"
> strings.replace_n({"f": "1", "foo":"2"}, "foo")
"1oo"
> strings.replace_n({"f": "1", "foo":"2"}, "foo")
"1oo"
> strings.replace_n({"f": "1", "foo":"2"}, "foo")
"1oo"
> strings.replace_n({"f": "1", "foo":"2"}, "foo")
"2"
```

## Steps to Reproduce the Problem

Run the snippet above many times, note the output.

## Additional Info

```interactive
$ opa version
Version: 0.25.0-dev
Build Commit: 467182247-dirty
Build Timestamp: 2020-10-25T11:50:25Z
Build Hostname: Stephans-MacBook-Pro.local
Go Version: go1.15.2
```

But I don't think that this has much to do with the OPA version used. It was introduced in v0.15.0.

### PR DESCRIPTION
**Pull request by srenatus**
Fixes #2822.

A quick check with the old and new implementation shows:

    $ for i in $(seq 0 1000); do ./opa_master eval --format=pretty 'strings.replace_n({"f":"x", "foo": "xxx"}, "foobar")' | grep -v xoobar; done | wc -l
         127
    $ for i in $(seq 0 1000); do ./opa_wip eval --format=pretty 'strings.replace_n({"f":"x", "foo": "xxx"}, "foobar")' | grep -v xoobar; done | wc -l
           0

So the non-determinism should be dealt with.

~Still planning to add test cases to pin down the current behaviour.~ ✔️

### CODE REVIEW

File: test/cases/testdata/replacen/test-replacen-0374.yaml [hunk 1]
```diff
@@ -14,3 +14,31 @@ cases:
   want_result:
   - x:
     - This is &lt;b&gt;HTML&lt;/b&gt;!
+- data: {}
+  modules:
+  - |
+    package generated
```
Comment (srenatus) [inline test/cases/testdata/replacen/test-replacen-0374.yaml:20]: 💭 What's the standard for test cases? "generated" isn't true anymore here...

Comment (tsandall) [reply test/cases/testdata/replacen/test-replacen-0374.yaml:20]: I've been using `package test` in most places. We might be able to just migrate all of the existing test cases to use that. Here's the hello-world example that I've pointed people at: https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/helloworld/test-helloworld-1.yaml

File: topdown/strings.go [hunk 2]
```diff
@@ -234,27 +233,31 @@ func builtinReplace(a, b, c ast.Value) (ast.Value, error) {
...
 
 	s, err := builtins.StringOperand(b, 2)
 	if err != nil {
 		return nil, err
 	}
 
 	var oldnewArr []string
-	for k, v := range oldnewObj {
-		strVal, ok := v.(string)
+	err = oldnewObj.Iter(func(k *ast.Term, v *ast.Term) error {
```
Comment (tsandall) [inline topdown/strings.go:247]: The object implementation iterates in insertion order so there could still be different outputs with this change. What about getting the keys from the object, sorting them, and then iterating over the sorted key set?

Comment (srenatus) [reply topdown/strings.go:247]: 🤔 As in `strings.replace_n(o1, s)` could differ from `strings.replace_n(o2, s)` although o1 and o2 are the same in terms of key-val pairs? Yup, and that is unfortunate. I had only considered multiple evals of `strings.replace_n(o1, s)` which should now no longer have different outcomes.

Sorting them would guarantee stable output, of course, but it would also mean that there's no way for the user to control the replacement order. I think this comes down to the interface of `strings.replace_n` being unfortunate, it's using something unordered, so you cannot control the ordering, end of story.

Since nobody has ever been complaining about this, it's probably not a big deal, the common "foo -> bar, baz -> quz" usecase isn't affected.

tl;dr: I'll update this, sorting the keys, tomorrow. Controlling the replacement ordering is just not going to be supported. (Or stay unsupported, rather.)

Comment (srenatus) [reply topdown/strings.go:247]: Technically, we could go back to using `ast.JSON`, and then sort the keys; but I think I like `builtins.ObjectOperand` better. WDYT?

Comment (tsandall) [reply topdown/strings.go:247]: > tl;dr: I'll update this, sorting the keys, tomorrow. Controlling the replacement ordering is just not going to be supported. (Or stay unsupported, rather.)

Sounds good. We can add that in the future if someone requests it.

> Technically, we could go back to using ast.JSON, and then sort the keys; but I think I like builtins.ObjectOperand better. WDYT?

I'd probably just use the `ast` types. I'd call `Keys()` and then sort the result.

Comment (srenatus) [reply topdown/strings.go]: ✅  Updated

**Review by tsandall — DISMISSED** [review]
LGTM

**Comment by tsandall** [comment]
@srenatus can you squash the commits? Once that's done we can merge.

**Comment by srenatus** [comment]
@tsandall sure! ✔️
